core/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -288,6 +288,7 @@ public synchronized IndexShard createShard(int sShardId, ShardRouting routing) {
             indicesLifecycle.afterIndexShardCreated(indexShard);
             settingsService.addListener(indexShard);
             shards = newMapBuilder(shards).put(shardId.id(), indexShard).immutableMap();
+            indexServicesProvider.getIndexingMemoryController().forceCheck();
```

Comment:
Actually, why not simply start up as inactive, init the buffers to the INACTIVE values, and then the first indexing op that arrives will make it active like normal?

OK that's a good point, that concurrent recoveries are limited anyway, so the O(N^2) cost is paid over a long time.